src/toxcore/tcp/mod.rs

```diff
@@ -0,0 +1,307 @@
+pub fn create_client_handshake(client_pk: PublicKey,
+                           client_sk: SecretKey,
+                           server_pk: PublicKey)
+    -> Result<(secure::Session, PrecomputedKey, ClientHandshake), Error>
+{
+    let session = secure::Session::new();
+    let payload = HandshakePayload { session_pk: *session.pk(), session_nonce: *session.nonce() };
+
+    let mut serialized_payload = [0; handshake::PAYLOAD_SIZE];
+    let (serialized_payload, _) = payload.to_bytes((&mut serialized_payload, 0)).unwrap();
+
+    let common_key = encrypt_precompute(&server_pk, &client_sk);
+    let nonce = gen_nonce();
+    let encrypted_payload = encrypt_data_symmetric(&common_key, &nonce, &serialized_payload);
+
+    let handshake = ClientHandshake { pk: client_pk, nonce: nonce, payload: encrypted_payload };
+    Ok((session, common_key, handshake))
+}
+
+/// Handle received client handshake on the server side.
+/// Return secure::Channel, Client PK, server handshake
+pub fn handle_client_handshake(server_sk: SecretKey,
+                           client_handshake: ClientHandshake)
+    -> Result<(secure::Channel, PublicKey, ServerHandshake), Error>
+{
+    let common_key = encrypt_precompute(&client_handshake.pk, &server_sk);
+    let payload_bytes = decrypt_data_symmetric(&common_key, &client_handshake.nonce, &client_handshake.payload)
+        .map_err(
+            |_| Error::new(ErrorKind::Other, "Failed to decrypt handshake::Client payload")
+        )?;
+
+    let payload = HandshakePayload::from_bytes(&payload_bytes).to_full_result().unwrap();
+
+    let client_pk = payload.session_pk;
+    let client_nonce = payload.session_nonce;
+
+    let session = secure::Session::new();
+    let server_payload = HandshakePayload { session_pk: *session.pk(), session_nonce: *session.nonce() };
+
+    let mut serialized_payload = [0; handshake::PAYLOAD_SIZE];
+    let (serialized_payload, _) = server_payload.to_bytes((&mut serialized_payload, 0)).unwrap();
+
+    let nonce = gen_nonce();
+    let server_encrypted_payload = encrypt_data_symmetric(&common_key, &nonce, &serialized_payload);
+
+    let server_handshake = ServerHandshake { nonce: nonce, payload: server_encrypted_payload };
+    let channel = secure::Channel::new(session, &client_pk, &client_nonce);
+    Ok((channel, client_handshake.pk, server_handshake))
+}
+
+/// Handle received server handshake on the client side.
+pub fn handle_server_handshake(common_key: PrecomputedKey,
+                           client_session: secure::Session,
+                           server_handshake: ServerHandshake)
+   -> Result<secure::Channel, Error>
+{
+    let payload_bytes = decrypt_data_symmetric(&common_key, &server_handshake.nonce, &server_handshake.payload)
+        .map_err(
+            |_| Error::new(ErrorKind::Other, "Failed to decrypt handshake::Server payload")
+        )?;
+    let payload = HandshakePayload::from_bytes(&payload_bytes).to_full_result().unwrap();
```

Comment:
What if it can't be parsed?

Reply:
Actually it cannot in the network code. But it seems like I can create that packet in unit tests. Thanks!

Reply:
At least it's not obvious that decrypt_data_symmetric returns and server_handshake.payload contains enough data.